### ISSUE TITLE
iptables in istio-init fails under OpenShift 4

### ISSUE THREAD

**Issue by henryz-vineti** [istio/istio#13986]
**Bug description**
Bookinfo fails to start under OpenShift 4. istio-init throws;
```
+ iptables -t nat -N ISTIO_REDIRECT
iptables v1.4.21: can't initialize iptables table `nat': Table does not exist (do you need to insmod?)
Perhaps iptables or your kernel needs to be upgraded.
+ dump
+ iptables-save
+ ip6tables-save
```

I've modified the configmap for istio-sidecar-injector per other issues to include;
```
      securityContext:
        capabilities:
          add:
            - NET_ADMIN
        privileged: true
        runAsUser: 0
        runAsNonRoot: false
        procMount: Default
```

**Expected behavior**
Pods should start

**Steps to reproduce the bug**
Deploy istio operator (`oc apply -n istio-operator -f https://raw.githubusercontent.com/Maistra/istio-operator/maistra-0.10/deploy/servicemesh-operator.yaml`) following the standard guide. Then deploy bookinfo.

**Version (include the output of `istioctl version --remote` and `kubectl version`)**
```
[root@localhost ~]# istioctl version --remote
client version: version.BuildInfo{Version:"1.1.5", GitRevision:"9b6d31b74d1c0cc9358cc82d395b53f71393326b", User:"root", Host:"3e29fde4-6c3f-11e9-b00d-0a580a2c0205", GolangVersion:"go1.10.4", DockerHub:"docker.io/istio", BuildStatus:"Clean", GitTag:"1.1.4-10-g9b6d31b"}
citadel version: version.BuildInfo{Version:"0.10.0-1", GitRevision:"a998810d2048a4c4acf206613f9c789a0df7a5aa", User:"redhat", Host:"redhat", GolangVersion:"go1.11.4", DockerHub:"brew-pulp-docker01.web.prod.ext.phx2.redhat.com:8888/openshift-istio-tech-preview", BuildStatus:"Clean", GitTag:"unknown"}
egressgateway version: version.BuildInfo{Version:"0.10.0-1", GitRevision:"4408997686d86a157f006403acfd46508d166f8c", User:"redhat", Host:"redhat", GolangVersion:"go1.11.4", DockerHub:"brew-pulp-docker01.web.prod.ext.phx2.redhat.com:8888/openshift-istio-tech-preview", BuildStatus:"Clean", GitTag:"unknown"}
galley version: version.BuildInfo{Version:"0.10.0-1", GitRevision:"a998810d2048a4c4acf206613f9c789a0df7a5aa", User:"redhat", Host:"redhat", GolangVersion:"go1.11.4", DockerHub:"brew-pulp-docker01.web.prod.ext.phx2.redhat.com:8888/openshift-istio-tech-preview", BuildStatus:"Clean", GitTag:"unknown"}
ingressgateway version: version.BuildInfo{Version:"0.10.0-1", GitRevision:"4408997686d86a157f006403acfd46508d166f8c", User:"redhat", Host:"redhat", GolangVersion:"go1.11.4", DockerHub:"brew-pulp-docker01.web.prod.ext.phx2.redhat.com:8888/openshift-istio-tech-preview", BuildStatus:"Clean", GitTag:"unknown"}
pilot version: version.BuildInfo{Version:"0.10.0-1", GitRevision:"a998810d2048a4c4acf206613f9c789a0df7a5aa", User:"redhat", Host:"redhat", GolangVersion:"go1.11.4", DockerHub:"brew-pulp-docker01.web.prod.ext.phx2.redhat.com:8888/openshift-istio-tech-preview", BuildStatus:"Clean", GitTag:"unknown"}
policy version: version.BuildInfo{Version:"0.10.0-1", GitRevision:"a998810d2048a4c4acf206613f9c789a0df7a5aa", User:"redhat", Host:"redhat", GolangVersion:"go1.11.4", DockerHub:"brew-pulp-docker01.web.prod.ext.phx2.redhat.com:8888/openshift-istio-tech-preview", BuildStatus:"Clean", GitTag:"unknown"}
sidecar-injector version: version.BuildInfo{Version:"0.10.0-1", GitRevision:"a998810d2048a4c4acf206613f9c789a0df7a5aa", User:"redhat", Host:"redhat", GolangVersion:"go1.11.4", DockerHub:"brew-pulp-docker01.web.prod.ext.phx2.redhat.com:8888/openshift-istio-tech-preview", BuildStatus:"Clean", GitTag:"unknown"}
telemetry version: version.BuildInfo{Version:"0.10.0-1", GitRevision:"a998810d2048a4c4acf206613f9c789a0df7a5aa", User:"redhat", Host:"redhat", GolangVersion:"go1.11.4", DockerHub:"brew-pulp-docker01.web.prod.ext.phx2.redhat.com:8888/openshift-istio-tech-preview", BuildStatus:"Clean", GitTag:"unknown"}
[root@localhost ~]# oc version
Client Version: version.Info{Major:"4", Minor:"1+", GitVersion:"v4.1.0-201905062232+ef90487-dirty", GitCommit:"ef90487", GitTreeState:"dirty", BuildDate:"2019-05-07T03:33:12Z", GoVersion:"go1.11.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"13+", GitVersion:"v1.13.4+34f13a8", GitCommit:"34f13a8", GitTreeState:"clean", BuildDate:"2019-05-09T21:13:37Z", GoVersion:"go1.11.8", Compiler:"gc", Platform:"linux/amd64"}
```

**How was Istio installed?**
Operator

**Environment where bug was observed (cloud vendor, OS, etc)**
OpenShift 4.0 rc1 in AWS, stock install

**Affected product area (please put an X in all that apply)**

[ ] Configuration Infrastructure
[ ] Docs
[X] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience

Additionally, please consider attaching a [cluster state archive](http://istio.io/help/bugs/#generating-a-cluster-state-archive) by attaching
the dump file to this issue.

**Comment by esnible** [comment]
Could this be a duplicate of https://github.com/istio/istio/issues/12948 and https://github.com/istio/istio/issues/4500 ?

**Comment by knrc** [comment]
@esnible No, they are not the same.  The other issues are related to the need for privileged containers which the Maistra distribution handles whereas this is a bug in the OCP 4.1 rc1 release as they have switched to nftables on the host and not correctly set up the iptables for the containers.

The OCP team expect iptables to continue working, see https://issues.jboss.org/browse/MAISTRA-410 for a workaround using a daemonset until this is fixed.

**Comment by jagusztinl** [comment]
The workaround not working, mentioned by knrc :-(

Before daemonset from upstream istio 0.12:

+ iptables -t nat -N ISTIO_REDIRECT
iptables v1.6.0: can't initialize iptables table `nat': Table does not exist (do you need to insmod?)
Perhaps iptables or your kernel needs to be upgraded.

After  installing the daemonset :
+ iptables -t nat -A ISTIO_REDIRECT -p tcp -j REDIRECT --to-port 15001
iptables: No chain/target/match by that name.

Not able to demo to our customer: Openshift 4.1 with Istio :-(

**Comment by sdake** [comment]
@knrc can you get this issue on the right person's plate?

Cheers
-steve

**Comment by knrc** [comment]
@sdake The issue is with the istio-init container, the community one expects iptables to be present on the host whereas OpenShift 4 uses nftables.

On the maistra side we have two init-containers, one with the older iptables utility (used with iptables hosts) and a second with the newer iptables utility (used with nftables hosts) however we are now switching completely over to CNI since the utility will come from the host itself.

I've raised this in a few meetings, especially when the topic of iptables/nftables arises, as we will start to see this more often as other kubernetes distributions also move to nftables.

**Comment by brian-avery** [comment]
The documentation has been fixed in #4905 to document that CNI must be used instead of proxy-init on OpenShift 4

**Comment by dvaldivia** [comment]
@brian-avery https://github.com/istio/istio/pull/4905 points to a PR that removes a couple of un-related lines imo, do you remember which was teh original issue you were referring to?
